File: Modules/ioblock/ioblock.cpp
/* -------------------------------------------
 * Blockade Modules - ioblock.cpp
 * -------------------------------------------
 * Implements a very hacky, but very useful
 * method for blocking reading and writing to
 * stdin and stdout, to prevent bots from
 * intefering with the arbiter-module stream.
 * Standard output is redirected to standard
 * error, so that logging output can still be
 * read.
 * -------------------------------------------
 * Author:  James Payor (github.com/amoebius)
 * -------------------------------------------
 */


#include "ioblock.hpp"
#include <unistd.h>
#include <fcntl.h>
#include <cstdio>

namespace ioblock {

	int infd, outfd, devnull;
	bool initialised = false, blocked;

	void init() {
		if(initialised) return;
		initialised = true;

		infd = dup(STDIN_FILENO);
		outfd = dup(STDOUT_FILENO);
		devnull = open("/dev/null", O_RDONLY);
		blocked = false;
	}

	void clean() {
		if(!initialised) return;
		unblock();

		close(infd);
		close(outfd);
		close(devnull);
		initialised = false;
	}

	void block() {
		if(blocked) return;
		blocked = true;

		fflush(stdout);
		dup2(devnull, STDIN_FILENO);
		dup2(STDERR_FILENO, STDOUT_FILENO);
	}

	void unblock() {
		if(!blocked) return;
		blocked = false;

		fflush(stdout);
		dup2(infd, STDIN_FILENO);
		dup2(outfd, STDOUT_FILENO);
	}

}; // namespace ioblock

File: Modules/ioblock/ioblock.py
import os
import sys

STDIN_FILENO = 0
STDOUT_FILENO = 1
STDERR_FILENO = 2

infd = os.dup(STDIN_FILENO)
outfd = os.dup(STDOUT_FILENO)
devnull = os.open("/dev/null", os.O_RDONLY)


blocked = False

def block():
	global blocked
	if blocked: return
	blocked = True

	sys.stdout.flush()
	os.dup2(devnull, STDIN_FILENO)
	os.dup2(STDERR_FILENO, STDOUT_FILENO)

def unblock():
	global blocked
	if not blocked: return
	blocked = False

	sys.stdout.flush()
	os.dup2(infd, STDIN_FILENO)
	os.dup2(outfd, STDOUT_FILENO)


class blocker_class(object):

	def __enter__(self):
		block()
		return blocker

	def __exit__(self, object_type, exception, traceback):
		unblock()
		return exception is None

blocker = blocker_class()
